specs/_features/eip6110/beacon-chain.md

```diff
@@ -0,0 +1,413 @@
+    # Finality
+    justification_bits: Bitvector[JUSTIFICATION_BITS_LENGTH]  # Bit set for every recent justified epoch
+    previous_justified_checkpoint: Checkpoint
+    current_justified_checkpoint: Checkpoint
+    finalized_checkpoint: Checkpoint
+    # Inactivity
+    inactivity_scores: List[uint64, VALIDATOR_REGISTRY_LIMIT]
+    # Sync
+    current_sync_committee: SyncCommittee
+    next_sync_committee: SyncCommittee
+    # Execution
+    latest_execution_payload_header: ExecutionPayloadHeader
+    # Withdrawals
+    next_withdrawal_index: WithdrawalIndex
+    next_withdrawal_validator_index: ValidatorIndex
+    # Deep history valid from Capella onwards
+    historical_summaries: List[HistoricalSummary, HISTORICAL_ROOTS_LIMIT]
+    # EIP-6110
+    deposit_receipts_start_index: uint64
+```
+
+## Beacon chain state transition function
+
+### Block processing
+
+```python
+def process_block(state: BeaconState, block: BeaconBlock) -> None:
+    process_block_header(state, block)
+    if is_execution_enabled(state, block.body):
+        process_withdrawals(state, block.body.execution_payload)
+        process_execution_payload(state, block.body.execution_payload, EXECUTION_ENGINE)  # [Modified in EIP-6110]
+    process_randao(state, block.body)
+    process_eth1_data(state, block.body)
+    process_operations(state, block.body)  # [Modified in EIP-6110]
+    process_sync_aggregate(state, block.body.sync_aggregate)
+```
+
+#### Modified `process_operations`
+
+*Note*: The function `process_operations` is modified to process `DepositReceipt` operations included in the payload.
+
+```python
+def process_operations(state: BeaconState, body: BeaconBlockBody) -> None:
+    # [New in EIP-6110]
+    # Prevent potential underflow introduced by mixing two deposit processing flows
+    if state.eth1_data.deposit_count > state.eth1_deposit_index:
+        assert len(body.deposits) == min(MAX_DEPOSITS, state.eth1_data.deposit_count - state.eth1_deposit_index)
+    else:
+        assert len(body.deposits) == 0
+
+    def for_ops(operations: Sequence[Any], fn: Callable[[BeaconState, Any], None]) -> None:
+        for operation in operations:
+            fn(state, operation)
+
+    for_ops(body.proposer_slashings, process_proposer_slashing)
+    for_ops(body.attester_slashings, process_attester_slashing)
+    for_ops(body.attestations, process_attestation)
+    for_ops(body.deposits, process_deposit)  # [Modified in EIP-6110]
+    for_ops(body.voluntary_exits, process_voluntary_exit)
+    for_ops(body.bls_to_execution_changes, process_bls_to_execution_change)
+
+    # [New in EIP-6110]
+    if is_execution_enabled(state, body):
+        for_ops(body.execution_payload.deposit_receipts, process_deposit_receipt)
+```
+
+#### New `get_validator_from_deposit_data`
+
+```python
+def get_validator_from_deposit_data(pubkey: BLSPubkey, withdrawal_credentials: Bytes32, amount: uint64) -> Validator:
```

Comment:
We could have `get_validator_from_deposit()` (defined in phase0) call this helper (and define this helper in phase 0 instead of here) for a little bit of code deduplication